stdlib/public/runtime/Reflection.mm

```diff
@@ -403,27 +403,47 @@ static bool loadSpecialReferenceStorage(HeapObject *owner,
                                         OpaqueValue *fieldData,
                                         const FieldType fieldType,
                                         Mirror *outMirror) {
-  // TODO: Switch over the other kinds of reference storage here:
-  // - unowned(safe)
-  // - unowned(unsafe)
-  // - class-bound existentials
-  // - Optional existential types
-  // This will require a change in the two low flag bits in the field type
-  // returned from the field type accessor generated by IRGen.
-
   // isWeak() implies a reference type via Sema.
   if (!fieldType.isWeak())
     return false;
 
+  auto type = fieldType.getType();
+  assert(type->getKind() == MetadataKind::Optional);
+
   auto weakField = reinterpret_cast<WeakReference *>(fieldData);
   auto strongValue = swift_unknownWeakLoadStrong(weakField);
-  fieldData = reinterpret_cast<OpaqueValue *>(&strongValue);
+
+  // Now that we have a strong reference, we need to create a temporary buffer
+  // from which to copy the whole value, which might be a native class-bound
+  // existential, which means we also need to copy n witness tables, for
+  // however many protocols are in the protocol composition. For example, if we
+  // are copying a:
+  // weak var myWeakProperty : (Protocol1 & Protocol2)?
+  // then we need to copy three pointesr:
+  // - the instance
+  // - the witness table for Protocol1
+  // - the witness table for Protocol2
+
+  // First create a temporary existential.
+  auto valueSize = type->getValueWitnesses()->getSize();
+  auto temporaryValue = static_cast<void*>(malloc(valueSize));
+
+  // Now copy the entire value out of the parent, which will include the
+  // witness tables.
+  memcpy(temporaryValue, reinterpret_cast<void *>(fieldData), valueSize);
```

Comment:
This is implicitly assuming that a weak reference is the same size as a pointer.  Let's avoid doing that.  You can cast temporaryValue to ClassExistentialContainer*; please add a WeakClassExistentialContainer, defined the same way except with a WeakReference instead of a void*, and memcpy to/from getWitnessTables().